Update js.js DOM text reinterpreted as HTML

Shivam7-1 opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

By using innerText, it will avoid the risk of HTML injection, as these properties automatically escape any HTML special characters in the provided text. This helps prevent cross-site scripting (XSS) vulnerabilities by treating the input as plain text rather than interpreted HTML.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#38767: Update js.js DOM text reinterpreted as HTML

Shivam7-1 committed:
Update js.js

Shivam7-1 commented:
@magento run all tests

Shivam7-1 commented:
@magento run all tests

engcom-Charlie commented:
@magento create issue

Shivam7-1 commented:
@magento run all tests

Shivam7-1 commented:
Hii @engcom-Charlie and @engcom-Hotel Could You Please Review This PR 
As Soon As Possible Because its pending for Review since many months 

Thanks & Regards

Shivam7-1 commented:
Hii @engcom-Charlie , @engcom-Dash @engcom-Bravo and @engcom-Hotel Could You Please Review This PR
As Soon As Possible Because its pending for Review since many months

Thanks & Regards

engcom-Hotel committed:
Merge branch '2.4-develop' into patch-1

engcom-Hotel commented:
@magento run all tests

engcom-Hotel reviewed (commented):
Hello @Shivam7-1,

Please let us know the steps to reproduce the actual issue for this PR that has been created.

Thanks

Shivam7-1 commented:
Hii @engcom-Hotel Thanks For Reviewing Here innerText is indeed Preferable over here instead of innerHtml which makes code more safer 
Eg https://www.freecodecamp.org/news/innerhtml-vs-innertext-vs-textcontent/

Shivam7-1 commented:
Hii @engcom-Hotel  Could You Please Review This again 
Thanks

engcom-Hotel commented:
Sure, @Shivam7-1, we will review it. But before that, could you please provide us with the manual testing scenario for this fix? It will help us proceed further with this PR.

Thanks

Shivam7-1 commented:
Hi @engcom-Hotel 

Thank you for reviewing the changes. As per your request
here is the manual testing scenario for the fix:

**Testing Scenario for Fix:**

1. **Initial State:**
   - Ensure the variable `valueError` is set to both `true` and `false` to cover both branches of the conditional statement.
   
2. **When `valueError` is `false`:**
   - Verify that the `this.advice.innerHTML` properly replaces the placeholder `%1` with `countDaysInMonth` in the `error` string. Since we replaced `innerHTML` with `innerText` in the updated code, ensure that the HTML content (if any) is rendered correctly as plain text, not as HTML tags.

3. **When `valueError` is `true`:**
   - Ensure the error text modification happens as expected and is set using `innerText` for the updated code. This will ensure that no HTML elements are unintentionally injected or rendered.

Why the change?
- I replaced `innerHTML` with `innerText` because `innerText` is generally safer to use in this context. Using `innerHTML` can sometimes expose the app to potential XSS (Cross-Site Scripting) vulnerabilities if any of the error messages contain user-inputted data or HTML tags. Since we are only displaying plain text without any HTML structure, `innerText` is more appropriate.

Let me know if you need any further clarification.

Thanks!

engcom-Hotel commented:
Hello @Shivam7-1,

Thanks for the reply!

Please elaborate on the steps for following the below step from frontend or admin panel:

> **Initial State:**
> 
> * Ensure the variable `valueError` is set to both `true` and `false` to cover both branches of the conditional statement.

Thanks

Shivam7-1 commented:
Hii @engcom-Hotel  Thanks For Reviewing again 
To clarify the testing steps for the initial state of the `valueError` variable here is how you can trigger and test both branches of the conditional statement from the frontend or admin panel:

 Testing for `valueError = false`:

1. **Go to the frontend or admin panel where the error message is displayed** (e.g., the page where the `advice` message appears).
2. Ensure that any conditions for setting `valueError` to `false` are met. This could be a situation where an action or validation fails, but the error is not critical, so `valueError` is set to `false`.
3. **Observe the error message in the UI**. Ensure that the error message uses `innerText` to display content. Since `innerText` doesn't parse HTML, any HTML tags (like `<b>`, `<i>`, etc.) should appear as plain text and not be rendered as HTML.
   
Testing for `valueError = true`:

1. **Trigger a scenario where `valueError` is set to `true`**. This could be a validation failure, form submission error, or another scenario where the error is critical.
2. Once the error is triggered, **check that the error message is displayed using `innerText`**, which will ensure no HTML content (like tags) is rendered, providing protection against XSS attacks.

 Why this change is needed:

By ensuring that both `valueError = true` and `valueError = false` are covered, we confirm that the switch from `innerHTML` to `innerText` correctly displays plain text content without the risk of rendering unintended HTML or executing scripts.

Let me know if you need further details or specific test cases for the frontend or admin panel.

Thanks!

Shivam7-1 commented:
Hii @engcom-Hotel @engcom-Bravo  Thanks For Reviewing PR
Is there anything Else is required from my side to get PR merge ?
Regards

engcom-Hotel commented:
Hello @Shivam7-1,

At this stage, no action is required from your side. The PR is currently in the testing bucket and will be progressed based on priority. We appreciate your patience.

Thanks

Shivam7-1 commented:
Hii @engcom-Hotel Thanks For Reply Could You Ping anyone from Team for Testing 
So it would get fast-track in Process as PR is much older 

Thanks

Shivam7-1 commented:
Hii @engcom-Hotel  @engcom-Charlie @engcom-Bravo  Could you Please ping tester here for tasting and get merge this PR 

So it will get merge Soon 
Thanks

engcom-Charlie commented:
Hi @Shivam7-1,

Thank you for your contribution!

Currently team is working on other priority tasks. We will pick this PR for further activities as per the priority.

Thank you!

Shivam7-1 commented:
Hii @engcom-Hotel Thanks For Response 
Is there any Update on this PR Testing?

Shivam7-1 commented:
Hii @engcom-Charlie have you assign this PR for Testing?

Shivam7-1 commented:
Hii @engcom-Hotel Thanks For Reply Could You Ping anyone from Team for Testing
So it would get fast-track in Process as PR is much older

Thanks

engcom-Charlie committed:
Merge branch '2.4-develop' into patch-1

Shivam7-1 commented:
Hii @engcom-Charlie Is there any update on above test ?
Thanks

engcom-Charlie commented:
Hi @Shivam7-1,

Thanks for the collaboration & contribution!

### :heavy_check_mark: QA Passed


**Preconditions:**

- Install fresh Magento **`2.4-develop`**
- Modify validation error message in lib/web/mage/validation.js eg. 
<img width="858" alt="image" src="https://github.com/user-attachments/assets/0eb33e2f-bd43-4b3b-a269-623784eb6da7" />


**Steps to reproduce**

- Go to Storefront
- Click on Create Account
- Add wrong email id in Email filed and observe the error message
- On PR branch, only the plain text is getting displayed escaping the HTML special characters because of using innerText property.



**Before:** :heavy_multiplication_x: 

<img width="724" alt="image" src="https://github.com/user-attachments/assets/e897f866-9f82-47d0-b6bb-3499969837ae" />








**After:** :heavy_check_mark: 

<img width="705" alt="image" src="https://github.com/user-attachments/assets/38ffe467-347d-401b-9f2b-b8c466865ef6" />




Builds are failed. Hence, moving this PR to **Extended Testing**.

Thanks.

engcom-Charlie committed:
Merge branch '2.4-develop' into patch-1

engcom-Charlie commented:
@magento run all tests

Shivam7-1 committed:
Merge branch '2.4-develop' into patch-1

Shivam7-1 commented:
Hii @engcom-Charlie Is there any update on above test ?
Thanks

engcom-Charlie commented:
@magento run all tests

engcom-Charlie commented:
Hi @Shivam7-1,

As mentioned [here](https://github.com/magento/magento2/pull/38756#issuecomment-2591772133), we are looking into the build test failures, we will let you know if we need your input in any case.

Thank you!

engcom-Charlie commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests

engcom-Charlie commented:
Some of the Functional CE test failures are not consistent in recent 2 builds and the consistent once are the known issues/

**Run 1:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38756/fe504e355935f9594267871c065d0d2a/Functional/allure-report-ce/index.html#categories/b687f15aea026e613d3b0ad4a593aa22/d8b00804567a96df/

<img width="1042" alt="image" src="https://github.com/user-attachments/assets/b95fde41-447e-404e-8a99-3cb5d85ec3f3" />


**Run 2:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38756/ece35a07807cd440af96f1003c0172bd/Functional/allure-report-ce/index.html#categories/7ea7ba6098fd5dba93af4c6a9157469a/e3aa1c60fcbde546/

<img width="1129" alt="image" src="https://github.com/user-attachments/assets/176ca722-9bd8-43b4-ae23-4faa1b96de46" />


Known issues:
[ACQE-7464](https://jira.corp.adobe.com/browse/ACQE-7464): VerifyDisableDownloadableProductSamplesAreNotAccessibleTest
[ACQE-7465](https://jira.corp.adobe.com/browse/ACQE-7465): VerifyOutOfStockDownloadableProductSamplesAreAccessibleTest

engcom-Charlie commented:
Some of the Functional B2B test failures are not consistent in recent 2 builds and the consistent once are the known issues/

**Run 1:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38756/46c769a1d661f97913ad1dd6b6e26a2e/Functional/allure-report-b2b/index.html#categories

<img width="1136" alt="image" src="https://github.com/user-attachments/assets/7afebc9c-0d82-4ab5-b805-79f48da6b4fb" />



**Run 2:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38756/18eb80cadc7d800785fe4b447766aa6f/Functional/allure-report-b2b/index.html#categories/12252aa641d411a980bdd869ac7bd814/a76b73840d85896a/

<img width="1135" alt="image" src="https://github.com/user-attachments/assets/49aa46d4-484e-48fd-b79d-a7d8a2846e19" />



**Known issues:**
[ACQE-7464](https://jira.corp.adobe.com/browse/ACQE-7464): VerifyDisableDownloadableProductSamplesAreNotAccessibleTest
[ACQE-7465](https://jira.corp.adobe.com/browse/ACQE-7465): VerifyOutOfStockDownloadableProductSamplesAreAccessibleTest
[ACQE-7387](https://jira.corp.adobe.com/browse/ACQE-7387): EndToEndB2CLoggedInUserTest
[AC-13447](https://jira.corp.adobe.com/browse/AC-13447): StorefrontQuoteCanBeRenamedUntilLockedTest
[ACQE-7460](https://jira.corp.adobe.com/browse/ACQE-7460): CheckingOfPreviewingRegistryUpdateEmailTest

engcom-Charlie commented:
Some of the Functional EE test failures are not consistent in recent 2 builds and the consistent once are the known issues/

**Run 1:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38756/a260929a4a7d67ddcca4e388171f614b/Functional/allure-report-ee/index.html#categories

<img width="1144" alt="image" src="https://github.com/user-attachments/assets/b5d76082-0203-4cad-8c14-523f88de2b35" />


**Run 2:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38756/89b42d7c7fc9fecb7cda3e586bea38e9/Functional/allure-report-ee/index.html#categories/25ecddfde592b027d2bf7a7bbc5e25ba/5890297edd7bb579/

<img width="1133" alt="image" src="https://github.com/user-attachments/assets/8ccad186-be56-4326-9f07-9c1329fc55a4" />




**Known issues:**
[ACQE-7464](https://jira.corp.adobe.com/browse/ACQE-7464): VerifyDisableDownloadableProductSamplesAreNotAccessibleTest
[ACQE-7465](https://jira.corp.adobe.com/browse/ACQE-7465): VerifyOutOfStockDownloadableProductSamplesAreAccessibleTest
[ACQE-7460](https://jira.corp.adobe.com/browse/ACQE-7460): CheckingOfPreviewingRegistryUpdateEmailTest

engcom-Charlie commented:
The Integration test failures are consistent in recent 2 builds and they are the known issues. Neither they are failing because of this PR nor part of this PR. Hence moving this PR to Merge in Progress.

**Run 1:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38756/51bd846f8120ac7f2afe87573e3e7fab/Integration/console-error-logs.html

<img width="1384" alt="image" src="https://github.com/user-attachments/assets/f805fe6a-f972-4c8b-8f58-bd7a980fb417" />



**Run 2:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38756/0adf949ed4088d09b2187c6c4708e9b8/Integration/console-error-logs.html

<img width="1390" alt="image" src="https://github.com/user-attachments/assets/8fdb70f7-3873-4a20-b7e1-2e23b718b299" />



**Known issues:**
[ACP2E-3531](https://jira.corp.adobe.com/browse/ACP2E-3531)
[AC-13688](https://jira.corp.adobe.com/browse/AC-13688)

Shivam7-1 commented:
Hii @engcom-Charlie just for confirmation is this PR is Ready to merge or anything is required from my side?
Thanks

Shivam7-1 commented:
Hii @engcom-Charlie @engcom-Hotel any update on this PR regarding merging related ?
Thanks

Shivam7-1 commented:
Hii @engcom-Charlie @engcom-Hotel @engcom-Dash any update on this PR regarding merging related ?
Thanks

engcom-Charlie commented:
Hi @Shivam7-1, As mentioned [here](https://github.com/magento/magento2/pull/38756#issuecomment-2594384187), we will let you know if anything is required from your side. We are working on priority work now, will take care of this PR's merging activity as per the priority.
